### PR TITLE
Malleable Blocks Part 2

### PR DESCRIPTION
**Pull request by Voxelot**
Followup PR to #180 to address some feedback there.

Also found some issues with how block id's were handled, so I changed the execution flow a bit to only generate a block id after the MST is finalized.

### CODE REVIEW

File: fuel-core/src/executor.rs [hunk 1]
```diff
@@ -103,20 +105,23 @@ impl Executor {
 
             // update block commitment
             let tx_fee = self.total_fee_paid(tx, vm_result.receipts())?;
-            // TODO: use SMT instead of this manual approach
+            // TODO: use merkle-sum-tree instead of this manual hashing approach
```
Comment (adlerjohn) [inline fuel-core/src/executor.rs:108]: This should just be a plain binary Merkle tree. I will update the block header specs soon for this.

Comment (Voxelot) [reply fuel-core/src/executor.rs:108]: Ah ok, well that should be an easy change.

File: fuel-core/src/model/fuel_block.rs [hunk 2]
```diff
@@ -69,72 +67,84 @@ impl BlockHeight {
...
-}
-
 #[derive(Clone, Debug, Deserialize, Serialize)]
-pub struct FuelBlockHeaders {
-    pub fuel_height: BlockHeight,
+pub struct FuelBlockHeader {
+    /// Fuel block height.
+    pub height: BlockHeight,
+    /// Ethereum block number.
+    pub number: BlockHeight,
```
Comment (rakita) [inline fuel-core/src/model/fuel_block.rs:75]: This is not something we know at moment of block creation, and we cant predict at what exact eth block is fuel block going to be commited. But it makes sense to be included as a field as it is going to be updated when fuel blocks get commited over eth. I would put this as `Option` or maybe zeroed value, and it needs to be removed from `id` hash calculation.

File: fuel-core/src/model/fuel_block.rs [hunk 3]
```diff
@@ -69,72 +67,84 @@ impl BlockHeight {
...
+    /// Block header hash of the previous block.
+    pub prev_hash: Bytes32,
```
Comment (rakita) [inline fuel-core/src/model/fuel_block.rs:77]: usualy called `parent` or `parent_hash`, but still `prev_hash` is fine name

File: fuel-core/src/model/fuel_block.rs [hunk 4]
```diff
@@ -69,72 +67,84 @@ impl BlockHeight {
...
+    /// Merkle root of all previous block header hashes.
+    pub prev_root: Bytes32,
```
Comment (rakita) [inline fuel-core/src/model/fuel_block.rs:79]: Not exactly sure why is this needed. Shouldn't `prev_hash` be enought to connect the chain?

File: fuel-core/src/executor.rs [hunk 5]
```diff
@@ -169,51 +169,72 @@ impl Executor {
...
             }
             ExecutionMode::Validation => {
-                if block.headers.transactions_commitment != commitment {
+                if block.header.transactions_root != txs_root {
                     return Err(Error::InvalidBlockCommitment);
                 }
             }
         }
+        // set the coinbase amount
+        block.header.coinbase = coinbase;
```
Comment (rakita) [inline fuel-core/src/executor.rs:216]: This probably still needs to be defines, but i think Nick mentioned long ago that fuel is going to be deflationary so in that sense `coinbase` needs to be increase by some fixed amount to reward validator and we will need to check under `ExecutionMode::Validation` if that amount is okay or not.

Comment (Voxelot) [reply fuel-core/src/executor.rs:216]: Good catch

File: fuel-core/src/model/fuel_block.rs (hunk 2 above)
Comment (Voxelot) [reply fuel-core/src/model/fuel_block.rs:75]: It seems hard to avoid non-deterministic behavior between validators if we don't commit to the ethereum global state as of this block. The block producer should be able to pick an ethereum block to claim that they are including all the ethereum events up to that point.

cc @adlerjohn

File: fuel-core/src/executor.rs (hunk 5 above)
Comment (Voxelot) [reply fuel-core/src/executor.rs:216]: Although it may still be useful to capture the total fees of the block, even if it doesn't all go directly to the validator and some get burned.

File: fuel-core/src/model/fuel_block.rs (hunk 4 above)
Comment (adlerjohn) [reply fuel-core/src/model/fuel_block.rs:79]: The hash of the previous block header is sufficient, but if it's a MMR of all previous block header hashes then you can do log-sized proofs to past block headers rather than linear.

File: fuel-core/src/executor.rs (hunk 5 above)
Comment (adlerjohn) [reply fuel-core/src/executor.rs]: I've been thinking of having a third coinbase tx type that's 0-input, 1-output. That way there's an explicit TXO for the coinbase rather than some implicit state mutation like with Ethereum.

File: fuel-core/src/model/fuel_block.rs (hunk 4 above)
Comment (rakita) [reply fuel-core/src/model/fuel_block.rs:79]: I see, this is not something needed for client as we need to have all blocks in db for rebroadcasting, but for v2 contract?  But for even v2, wouldn't it be enought to have map of index->hash and there wouldn't be need MMR hashing and checking. Just Wondering, yes we have log-sized proof but where is it needed to have that reduction?

File: fuel-core/src/executor.rs (hunk 5 above)
Comment (rakita) [reply fuel-core/src/executor.rs]: > I've been thinking of having a third coinbase tx type that's 0-input, 1-output. That way there's an explicit TXO for the coinbase rather than some implicit state mutation like with Ethereum.

This is similar problem with token deposits, we need to make token out of thin air in fuel :). Maybe having additional Input that will cover those two cases?

Comment (adlerjohn) [reply fuel-core/src/executor.rs]: There's an issue to track deposit IDs here: https://github.com/FuelLabs/fuel-specs/issues/106

File: fuel-core/src/model/fuel_block.rs (hunk 4 above)
Comment (adlerjohn) [reply fuel-core/src/model/fuel_block.rs:79]: Storing every block header hash would mean a storage write to a new slot for every block. If we use an MMR, then we only need to keep track of the tip's header hash, which is much cheaper.

File: fuel-core/src/model/fuel_block.rs (hunk 2 above)
Comment (Voxelot) [reply fuel-core/src/model/fuel_block.rs:75]: I was thinking this would be L1 block number used for synchronizing deposits, not the L1 block that this block ended up being published in.

Comment (adlerjohn) [reply fuel-core/src/model/fuel_block.rs:75]: That's a possibility, yes!

Comment (adlerjohn) [reply fuel-core/src/model/fuel_block.rs:75]: I'm okay with leaving a field in the header for Ethereum block, but there should be a description here that there's some caveats, and linking to https://github.com/FuelLabs/fuel-specs/issues/220

File: fuel-core/src/model/fuel_block.rs (hunk 3 above)
Comment (Voxelot) [reply fuel-core/src/model/fuel_block.rs:77]: I like parent hash as well

File: fuel-core/src/executor.rs (hunk 5 above)
Comment (Voxelot) [reply fuel-core/src/executor.rs]: I've removed the coinbase field and will just `debug` log the fee amounts for now.

File: fuel-core/src/model/fuel_block.rs (hunk 4 above)
Comment (Voxelot) [reply fuel-core/src/model/fuel_block.rs:83]: Which of our MT implementations supports MMR?

Comment (adlerjohn) [reply fuel-core/src/model/fuel_block.rs:83]: The binary Merkle tree is actually an MMR 🤷

File: fuel-core/src/model/fuel_block.rs (hunk 2 above)
Comment (Voxelot) [reply fuel-core/src/model/fuel_block.rs:79]: updated the description of this field, lmk if that works for you.

Comment (adlerjohn) [reply fuel-core/src/model/fuel_block.rs:79]: Good enough!

**Review by adlerjohn — COMMENTED** [review]
No further comments from me, will defer to @rakita

File: fuel-core/src/model/fuel_block.rs (hunk 4 above)
Comment (Voxelot) [reply fuel-core/src/model/fuel_block.rs:83]: I think I'll defer this to a separate task

Comment (Voxelot) [reply fuel-core/src/model/fuel_block.rs:83]: https://github.com/FuelLabs/fuel-core/issues/192
